web_monitoring/db.py

```diff
@@ -126,7 +126,7 @@ class method :meth:`Client.from_env`.
     url : string, optional
         Default is ``https://api.monitoring.envirodatagov.org``.
     """
-    def __init__(self, email, password, url=DEFAULT_URL):
+    def __init__(self, email='', password='', url=DEFAULT_URL):
```

Comment:
@danielballan In case the user wants to use `from_env` later, there shouldn't be an error message when the user tries to create an object like this `obj1 = Client()'.

Reply:
I don’t think that’s right—`from_env` is a class method, so you would never use it “later;” it always creates a fresh instance of `Client`.

> there shouldn't be an error message when the user tries to create an object like this `obj1 = Client()`

Credentials should *always* be provided to any `Client` instance. You can’t access the DB without them (that update hasn’t been deployed yet, but will be soon). I‘m pretty sure these arguments are required on purpose.

Reply:
@Mr0grog if credentials should always be provided, then the option to use ‘from_env’ should not be there or it should be a part of the constructor.
Shouldn’t it?

Reply:
I think the intent here is to not make it part of the constructor so that what you are asking for is explicit. It also makes the contract of the API more clear: you can’t misunderstand that you must provide credentials—in some form—in order to create a client.

It also leaves open the opportunity to have a client *without* credentials (we just make them optional instead of required) when we eventually relax the login requirements on the DB side. You don’t have to worry about revising the constructor to no longer fill in default values from the environment and also fixing all code other people have written not in this project that might have depended on that behavior.

Reply:
Oh, I see. I suppose any user must be clear about the idea of creating a class with either the constructor or by using `from_env` if they're using this API. 
Though a class method should not be used later, it is possible to use it, and that has bothered me earlier as well.

Reply:
> It also makes the contract of the API more clear: you can’t misunderstand that you must provide credentials—in some form—in order to create a client.

Exactly.

> Though a class method should not be used later, it is possible to use it, and that has bothered me earlier as well.

That's a fair criticism of this pattern: the fact that it's possible to do:

```python
new_client_instance = client_instance.from_env()
```

is a bit weird. Nonetheless, I find that using class methods to implement "alternative" constructors is a fairly popular pattern. An alternative approach would be to implement this as a separate function, like `make_client_from_env()`.

Reply:
It is quite popular. I suppose we can add a little something about class methods in the docstring.